compiler/nilcheck.nim

```diff
@@ -0,0 +1,928 @@
+
+
+  of nkAsgn:
+    result = checkAsgn(n[0], n[1], ctx, map)
+  of nkVarSection:
+    result.map = map
+    for child in n:
+      aecho child.kind
+      result = checkAsgn(child[0], child[2], ctx, result.map)
+  of nkForStmt:
+    result = checkFor(n, ctx, map)
+  of nkCaseStmt:
+    result = checkCase(n, ctx, map)
+  of nkReturnStmt:
+    result = checkReturn(n, ctx, map)
+  of nkBracketExpr:
+    result = checkBracketExpr(n, ctx, map)
+  of nkTryStmt:
+    result = checkTry(n, ctx, map)
+  of nkWhileStmt:
+    result = checkWhile(n, ctx, map)
+  of nkNone..pred(nkSym), succ(nkSym)..nkNilLit, nkTypeSection, nkProcDef, nkConverterDef,
+      nkMethodDef, nkIteratorDef, nkMacroDef, nkTemplateDef, nkLambda, nkDo,
+      nkFuncDef, nkConstSection, nkConstDef, nkIncludeStmt, nkImportStmt,
+      nkExportStmt, nkPragma, nkCommentStmt, nkBreakState, nkTypeOfExpr:
+    
+    discard "don't follow this : same as varpartitions"
+    result = (Nil, map)
+  else:
+    var elementMap = map.copyMap()
+    var nilability = Safe
+    for element in n:
+      (nilability, elementMap) = check(element, ctx, elementMap)
+
+    result = (Nil, elementMap)
+
+  
+
+
+  
+  # echo map
+
+proc typeNilability(typ: PType): Nilability =
+  #if not typ.isNil:
+  #  echo ("type ", typ, " ", typ.flags, " ", typ.kind)
+  if typ.isNil: # TODO is it ok
+    Safe
+  elif tfNotNil in typ.flags:
+    Safe
+  elif typ.kind in {tyRef, tyCString, tyPtr, tyPointer}:
+    # 
+    # tyVar ? tyVarargs ? tySink ? tyLent ?
+    # TODO spec? tests? 
+    MaybeNil
+  else:
+    Safe
+
+
+proc toSymbolGraphs(partitions: Partitions): Table[Symbol, GraphIndex] =
+  for graphIndex, graph in partitions.graphs:
+    for element in graph.elements:
+      result[symbol(element)] = graphIndex
+
+proc checkNil*(s: PSym; body: PNode; conf: ConfigRef, partitions: Partitions) =
+  var map = newNilMap()
+  let line = s.ast.info.line
+  let fileIndex = s.ast.info.fileIndex.int
+  var filename = conf.m.fileInfos[fileIndex].fullPath.string
+
+  # echo toTextGraph(conf, partitions)
+  
+  # TODO
+  var context = NilCheckerContext(partitions: partitions, symbolGraphs: toSymbolGraphs(partitions), config: conf)
+  for i, child in s.typ.n.sons:
+    if i > 0:
+      if child.kind != nkSym:
+        continue
+      map.store(context.graph(child), typeNilability(child.typ), TArg, child.info, child)
+
+  map.store(context.graph(resultId), if not s.typ[0].isNil and s.typ[0].kind == tyRef: Nil else: Safe, TResult, s.ast.info)
+    
+  # echo "checking ", s.name.s, " ", filename
+
+  let res = check(body, context, map)
+  let resultGraphIndex = context.graph(resultId)
+  if res.nilability == Safe and (not res.map.history.hasKey(resultGraphIndex) or res.map.history[resultGraphIndex].len <= 1):
+    res.map.store(resultGraphIndex, Safe, TAssign, s.ast.info)
+  if not s.typ[0].isNil and s.typ[0].kind == tyRef and tfNotNil in s.typ[0].flags:
```

Comment:
What if it's not a ``tyRef`` but a tuple containing a ref?

Reply:
good point: I will think

Reply:
I am starting a commented out test for that: currently i think we should just warn for uninitialized exprs requiring non nilable and for assigning a nilable element to non-nilable (the second one seems to be sometimes happening already in the compiler)

Reply:
but i'll wait for tomorrow